src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+      const paramTSIds = timeseries[i++].map(({ id }) => id);
+      const pseudoParsedSTS = paramTSIds.map(id =>
+        STSReference([STSFilter('id', id), ...syntheticFilters])
+      );
+      return pseudoParsedSTS.map(ts => reverse(ts)).join(', ');
+    }
+  });
+  return flattenSumFunctions(exprWithSum);
+};
+
+export const parse = (s: string): STSQueryItem[] | STSFunction => {
+  return parseWith(new Parser(compiledGrammar), s)
+};
+
+export const generateAllPossiblePermutations = <T>(
+  arrayOfArrays: T[][],
+  lockIndices: number[] = []
+): T[][][] => {
+  let res: T[][][] = [];
+  for (let i = 0; i < arrayOfArrays.length; i++) {
+    const locked = lockIndices.includes(i);
+    const array = [...arrayOfArrays[i]];
+    if (res.length) {
+      const temp = [...res];
+      const temptemp = [];
+      for (const t of temp) {
+        if (locked) {
+          temptemp.push([...t, array]);
+        } else {
+          for (const item of array) {
+            temptemp.push([...t, [item]]);
+          }
+        }
+      }
+      res = [...temptemp];
+    } else {
+      if (locked) {
+        res = [[array]];
+      } else {
+        res = array.map(i => [[i]]);
+      }
+    }
+  }
+  return res;
+};
+
+export const convertExpressionToLabel = (
+  expression: string,
+  labelSrc: string,
+  tsMap: TSResponseMap
+) => {
+  const parsed = parse(expression);
+  const labelRes = reverse(parsed, item => {
+    if (isSTSReference(item)) {
+      const [{ value }] = getIdFilters(item);
+      const serie = tsMap[String(value)];
+      if (labelSrc) {
+        return getLabelWithInjectedProps(labelSrc, serie);
+      }
+      return `${serie.name || serie.externalId || serie.id}`;
+    }
+  });
+  return labelRes;
+};
+
+export const getLabelsForExpression = async (
+  expressions: string[],
+  labelSrc: string,
+  target,
+  connector: Connector
+) => {
+  const tsIds = getReferencedIdsInExpressions(expressions);
+  const tsUniqueIds = uniqBy(tsIds, unwrapId);
+  const referencedTS = await getTimeseries({ items: tsUniqueIds }, target, connector, false);
+  const tsMap = reduceTsToMap(referencedTS);
+  return expressions.map(expr => convertExpressionToLabel(expr, labelSrc, tsMap));
+};
+
+export const getReferencedTimeseries = (
+  route: STSQueryItem[] | STSFunction
+): StringMap[] => {
+  let idFilters: STSServerFilter[] = [];
+  walk(route, obj => {
+    if (isSTSReference(obj)) {
+      const ids = getIdFilters(obj);
+      idFilters = [...idFilters, ...ids];
+    }
+  });
+  return idFilters.map(({ path, value }) => ({
+    [path]: value as string,
+  }));
+};
+
+export const hasAggregates = (
+  expression: string
+): boolean => {
+  let hasAggregates = false;
+  walk(parse(expression), obj => {
+    if (isSTSReference(obj) && obj.query.some(isSTSAggregateFilter)) {
+      hasAggregates = true;
+    }
+  });
+  return hasAggregates
+};
+
+export function flattenServerQueryFilters(items: STSFilter[]): StringMap {
```

Comment:
mb we can have only function expression in a file (like function above)? Replace
```typescript
export function flattenServerQueryFilters(items: STSFilter[]): StringMap { ... }
```
with
```typescript
export const flattenServerQueryFilters = (items: STSFilter[]): StringMap => { ... }
```

Reply:
you mean arrow functions for all exported members?

Reply:
yep

Reply:
fixed